File: app/src/main/java/com/simo/recylerviewjsonexample/MainActivity.java
package com.simo.recylerviewjsonexample;

import android.os.Bundle;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MainActivity extends AppCompatActivity {

    private RecyclerView mRecyclerView;
    private ExampleAdapter mExampleAdapter;
    private ArrayList<ExampleItem> mExampleList;
    private RequestQueue mRequestQueue;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mRecyclerView = findViewById(R.id.recycler_view);
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        mExampleList = new ArrayList<>();

        mRequestQueue = Volley.newRequestQueue(this);
        parseJSON();

    }

    // parsing Json File

    private void parseJSON() {

        String url = "https://api.github.com/search/repositories?q=created:>"+getLastMonth(30)+"&sort=stars&order=desc";

        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, url, null,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        try {
                           // JSONArray jsonArray = response.getJSONArray("hits");
                            JSONArray jsonArray = response.getJSONArray("items");

                            for (int i = 0; i < jsonArray.length(); i++) {
                                JSONObject hit = jsonArray.getJSONObject(i);

                                String repoName = hit.getString("name");
                                String repoDescription = hit.getString("description");
                                String repoAvatarImg = hit.getJSONObject("owner").getString("avatar_url");
                                String repoOwner = hit.getJSONObject("owner").getString("login");
                                double stars = hit.getDouble("score");

                                mExampleList.add(new ExampleItem(repoAvatarImg,repoOwner,stars,repoDescription,repoName));
                            }

                            mExampleAdapter = new ExampleAdapter(MainActivity.this, mExampleList);
                            mRecyclerView.setAdapter(mExampleAdapter);

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                error.printStackTrace();
            }
        });

        mRequestQueue.add(request);
    }

    private String getLastMonth(int days)
    {
        // pick current system date
        GregorianCalendar gc = new GregorianCalendar();
        // find the current date - number of days
        gc.add(Calendar.DATE, -days);
        // set format for date
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
        // parse it like
        String resDate = dateFormat.format( gc.getTime());
        return  resDate;

    }
}

File: app/src/main/java/com/example/mobilecodingchallenge/Controller/MainActivity.kt
package com.example.mobilecodingchallenge

import android.icu.text.SimpleDateFormat
import android.icu.util.Calendar
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    lateinit var adapter: GitHubRepoRecycleAdapter
    var gitHubRepo: List<Items> = listOf()
    private var isServiceRunning = false
    private var page = 1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        adapter = GitHubRepoRecycleAdapter(this,gitHubRepo.toMutableList())
        gitHubRepoListView.adapter = adapter
        gitHubRepoListView.layoutManager = LinearLayoutManager(this)
        gitHubRepoListView.setHasFixedSize(true)

        loadData(getQueryString())

        gitHubRepoListView.addOnScrollListener(object: RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val linearLayoutManager = recyclerView
                    .layoutManager as LinearLayoutManager?

                if (!isServiceRunning && linearLayoutManager!!.itemCount <= linearLayoutManager.findLastVisibleItemPosition() + 2) {
                    page = page + 1
                    loadData(getQueryString(page))
                }
            }

        })
    }

    fun loadData(queryString: String) {
        progressBar.visibility = View.VISIBLE
        isServiceRunning = true
        WebService.callService(this,queryString,{response ->
            isServiceRunning = false
            progressBar.visibility = View.INVISIBLE
            if (gitHubRepo.count() == 0) {
                gitHubRepo = response.gitHubRepo
                adapter.setData(response.gitHubRepo)
            } else {
                adapter.addMoreData(response.gitHubRepo)
            }
        },{
            Toast.makeText(this,it,Toast.LENGTH_LONG).show()
            isServiceRunning = false
            progressBar.visibility = View.INVISIBLE
        })
    }

    fun getQueryString(page: Int = 1) : String {
        val date = Calendar.getInstance()
        date.add(Calendar.DATE, -1)

        val formatter = SimpleDateFormat("yyyy-MM-dd")
        val thirtyDaysBeforeTodayInString = formatter.format(date)

        var queryString = "q=created:%3E$thirtyDaysBeforeTodayInString&sort=stars&order=desc"
        if (page > 1) {
            queryString = "$queryString&page=$page"
        }
        return queryString
    }
}
